presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -556,43 +582,193 @@ private static NestedField nestedField(String name)
-                this.connectorSession = requireNonNull(connectorSession, "connectorSession is null");
+                this.connectorSession = connectorSession;
+                this.isPushDownSubfieldsFromLambdasEnabled = isPushDownSubfieldsFromLambdasEnabled;
+                this.functionAndTypeManager = functionAndTypeManager;
             }
 
             @Override
             public Void visitCall(CallExpression call, Context context)
             {
-                if (!functionResolution.isSubscriptFunction(call.getFunctionHandle())) {
-                    call.getArguments().forEach(argument -> argument.accept(this, context));
+                ComplexTypeFunctionDescriptor functionDescriptor = functionAndTypeManager.getFunctionMetadata(call.getFunctionHandle()).getDescriptor();
+                if (functionResolution.isSubscriptFunction(call.getFunctionHandle()) ||
+                        functionAndTypeManager.getFunctionMetadata(call.getFunctionHandle()).getName().equals(ELEMENT_AT)) {
+                    Optional<Subfield> subfield = toSubfield(call, functionResolution, expressionOptimizer, connectorSession, functionAndTypeManager);
+                    if (subfield.isPresent()) {
+                        if (context.isPruningLambdaSubfieldsPossible()) {
+                            addRequiredLambdaSubfields(context, subfield.get());
+                            return null;
+                        }
+                        context.subfields.add(subfield.get());
+                    }
+                    else {
+                        call.getArguments().forEach(argument -> argument.accept(this, context));
+                    }
                     return null;
                 }
+                if (isPushDownSubfieldsFromLambdasEnabled) {
+                    Optional<Set<Subfield>> lambdaSubfieldsOriginal = context.getLambdaSubfields();
+                    if ((functionDescriptor.isAccessingInputValues() && functionDescriptor.getLambdaDescriptors().isEmpty())
+                            || (functionDescriptor.isTransformingInput()) && !functionDescriptor.getOutputToInputTransformationFunction().isPresent()) {
+                        // If function internally accesses the input values we cannot prune any unaccessed lambda subfields since we do not know what subfields function accessed.
+                        // If function transforms the shape of the input map or array argument and the form of such transformation was not supplied in
+                        // getOutputToInputTransformationFunction() then we do not know how to attribute those subfields to the input argument. Therefore, we have to give up on
+                        // pruning unaccessed subfields and read all of them.
+                        context.giveUpOnCollectingLambdaSubfields();
+                    }
+                    if (!functionDescriptor.isReturningInputValues()) {
```

Comment:
we should check isReturningInputValues() above, and shouldn't need this if condition or the other places where we check for this case.

Reply:
removed `isReturningInputValues` in separate commit that addresses this comment.